### PR TITLE
optimize torchnet

### PR DESCRIPTION
**Pull request by qiuxin2012**
As pytorch's `backward()` is serialized per device, see [pytorch issue #18333](https://github.com/pytorch/pytorch/issues/18333) and [pytorch issue #17566](https://github.com/pytorch/pytorch/issues/17566) for details, the performance of default distributed training mode(multi models on each executor) will be very bad. In order to reach the best performance, TorchNet's training should only create one model in each executor, and use multi OMP threads to speedup the single model's training.

This pr provide a way to train pytorch model in single model + multi OMP threads mode, when user 
set ZOO_NUM_MKLTHREADS to all using `export ZOO_NUM_MKLTHREADS=all`.

### CODE REVIEW

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 1]
```diff
@@ -1083,7 +1083,23 @@ private[zoo] class InternalDistriOptimizer[T: ClassTag] (
...
     val nodeNumber = EngineRef.getNodeNumber()
-    val coresPerNode = EngineRef.getCoreNumber()
+
+    /**
+     * The best practice of torchnet's training is single model in each executor.
+     * And use multi OMP threads to speedup the single model's training.
+     * Currently, we only provide single model + multi OMP threads for torchnet model.
+     */
+    val torchNetOptimize = model.isInstanceOf[TorchNet] &&
+      (System.getenv("OMP_NUM_THREADS") != null)
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: In our script OMP_NUM_THREADS will always be set and won't be null?

Comment (qiuxin2012) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: In case user use zoo without our starting script...

Comment (qiuxin2012) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: init_spark_on_yarn allow user create zoo enviroment just `python`.

File: pyzoo/zoo/examples/pytorch/train/resnet_finetune/README.md [hunk 2]
```diff
@@ -49,6 +50,7 @@ You can easily use the following commands to run this example:
     hdfs dfs -put /tmp/zoo/dogs_cats dogs_cats 
     export HADOOP_CONF_DIR=[path to your hadoop conf directory]
     export ZOO_CONDA_NAME=[conda environment name you just prepared above]
+    export OMP_NUM_THREADS=4
```
Comment (jason-dai) [inline pyzoo/zoo/examples/pytorch/train/resnet_finetune/README.md]: We should try to do this programmably, instead of using script

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 3]
```diff
@@ -1083,7 +1083,20 @@ private[zoo] class InternalDistriOptimizer[T: ClassTag] (
...
+
+    /**
+     * The best practice of torchnet's training is single model in each executor.
+     * And use multi OMP threads to speedup the single model's training.
+     * Currently, we only provide single model + multi OMP threads for torchnet model.
+     */
+    val torchNetOptimize = model.isInstanceOf[TorchNet]
+    val modelPerExecutor = if (torchNetOptimize) {
+      require(EngineRef.getEngineType() != MklDnn, "torchnet shouldn't use MKLDNN engine.")
+      logger.info(s"torchnet will use ${EngineRef.getCoreNumber()} OMP threads.")
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: where do you set OMP_THREAD for Pytorch?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 4]
```diff
@@ -1125,9 +1138,16 @@ private[zoo] class InternalDistriOptimizer[T: ClassTag] (
...
       val modelsAndBroadcast = InternalOptimizerUtil.initThreadModels[T](
         trainingModel, distDataset, criterion, state,
-        Int.box(nodeNumber), Int.box(coresPerNode), Boolean.box(checkSingleton),
+        Int.box(nodeNumber), Int.box(modelPerExecutor), Boolean.box(checkSingleton),
         allReduceParameter, parameterSplits, validationMethods, optimMethods, parameterProcessors)
       cachedModels = modelsAndBroadcast._1
+      if (torchNetOptimize) {
+        val numOmpThread = EngineRef.getCoreNumber()
+        cachedModels.map{_ =>
+          EngineRef.getDefaultThreadPool().setPoolSize(numOmpThread)
```
Comment (qiuxin2012) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: We set OMP_THREAD here.

File: docs/docs/ProgrammingGuide/pytorch.md [hunk 5]
```diff
@@ -153,3 +153,8 @@ and we expects to see the output like:
...
 
+# Training Best Practise
+As pytorch's `backward()` is serialized per device, see [pytorch issue #18333](https://github.com/pytorch/pytorch/issues/18333) and [pytorch issue #17566](https://github.com/pytorch/pytorch/issues/17566) for detail, the performance of default distributed training mode(multi models on each executor) will be very bad. So to reach the best performance, torchnet's training should only create one model in each executor, and use multi OMP threads to speedup the single model's training.
+
+We recommend you use [Estimator](../APIGuide/PipelineAPI/estimator.md), [NNEstimator](../APIGuide/PipelineAPI/nnframes.md#nnestimator) or [NNClassifier](../APIGuide/PipelineAPI/nnframes.md#nnclassifier
+) for pytorch model training, when Estimator, NNEstimator and NNClassifier detect Torchnet model, they will optimize training using single model multi OMP threads automatically. Here is a [ResNet50 finetune example](https://github.com/qiuxin2012/analytics-zoo/blob/pytorch-multiexecutor/pyzoo/zoo/examples/pytorch/train/resnet_finetune/README.md).
```
Comment (hkvision) [inline docs/docs/ProgrammingGuide/pytorch.md]: uri of the example is in your own repo?

File: docs/docs/ProgrammingGuide/pytorch.md [hunk 6]
```diff
@@ -153,3 +153,8 @@ and we expects to see the output like:
 
 More Pytorch examples (ResNet, Lenet etc.) are available [here](https://github.com/intel-analytics/analytics-zoo/tree/master/pyzoo/zoo/examples/pytorch).
 
+# Training Best Practise
+As pytorch's `backward()` is serialized per device, see [pytorch issue #18333](https://github.com/pytorch/pytorch/issues/18333) and [pytorch issue #17566](https://github.com/pytorch/pytorch/issues/17566) for detail, the performance of default distributed training mode(multi models on each executor) will be very bad. So to reach the best performance, torchnet's training should only create one model in each executor, and use multi OMP threads to speedup the single model's training.
```
Comment (hkvision) [inline docs/docs/ProgrammingGuide/pytorch.md]: detail -> details
So -> In order
torchnet -> TorchNet

File: docker/zoo/start-notebook.sh [hunk 7]
```diff
@@ -45,12 +45,6 @@ fi
 if [ -z "${OMP_NUM_THREADS}" ]; then
     if [ -z "${ZOO_NUM_MKLTHREADS}" ]; then
         export OMP_NUM_THREADS=1
-    else
-        if [ `echo $ZOO_NUM_MKLTHREADS | tr '[A-Z]' '[a-z]'` == "all" ]; then
-            export OMP_NUM_THREADS=`nproc`
-        else
-            export OMP_NUM_THREADS=${ZOO_NUM_MKLTHREADS}
-        fi
     fi
```
Comment (jason-dai) [inline docker/zoo/start-notebook.sh]: we still need this to set `OMP_NUM_THREADS` on driver node

File: pyzoo/dev/prepare_env.sh [hunk 8]
```diff
@@ -45,12 +45,6 @@ fi
 if [ -z "${OMP_NUM_THREADS}" ]; then
     if [ -z "${ZOO_NUM_MKLTHREADS}" ]; then
         export OMP_NUM_THREADS=1
-    else
-        if [ `echo $ZOO_NUM_MKLTHREADS | tr '[A-Z]' '[a-z]'` == "all" ]; then
-            export OMP_NUM_THREADS=`nproc`
-        else
-            export OMP_NUM_THREADS=${ZOO_NUM_MKLTHREADS}
-        fi
     fi
```
Comment (jason-dai) [inline pyzoo/dev/prepare_env.sh]: we still need this to set `OMP_NUM_THREADS` on driver node

File: pyzoo/zoo/common/nncontext.py [hunk 9]
```diff
@@ -178,7 +178,7 @@ def init_env():
         omp_num_threads = os.environ["OMP_NUM_THREADS"]
     elif "ZOO_NUM_MKLTHREADS" in os.environ:
         if os.environ["ZOO_NUM_MKLTHREADS"].lower() == "all":
-            omp_num_threads = multiprocessing.cpu_count()
+            omp_num_threads = conf.get('spark.executor.cores', str(multiprocessing.cpu_count()))
```
Comment (jason-dai) [inline pyzoo/zoo/common/nncontext.py]: user may not set `spark.executor.cores`; I think you need to use the same logic in `Engine.init` to get the core number

Comment (qiuxin2012) [reply pyzoo/zoo/common/nncontext.py]: Engine is using `spark.executor.cores` to get core number.
Only spark local mode doesn't have `spark.executor.cores`, then we use all cpu in local machine.

Comment (jason-dai) [reply pyzoo/zoo/common/nncontext.py]: See:
https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala#L485 and https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala#L251

Comment (qiuxin2012) [reply pyzoo/zoo/common/nncontext.py]: In https://github.com/intel-analytics/BigDL/blob/master/spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/Engine.scala#L485, all cluster mode(spark standalone, yarn, mesos, k8s) are all using `val coreString = conf.get("spark.executor.cores", null)` to get core number. This is the same logic compared with `Engine.init`.
If `spark.executor.cores` is not set, it means we are using spark local or pure java, then we will use all cores in local machine. This is a little different from `Engine.init`.

File: pyzoo/zoo/common/nncontext.py [hunk 10]
```diff
@@ -174,50 +174,48 @@ def init_env():
...
         else:
             omp_num_threads = os.environ["ZOO_NUM_MKLTHREADS"]
+    elif "OMP_NUM_THREADS" in os.environ:
+        omp_num_threads = os.environ["OMP_NUM_THREADS"]
     if "KMP_BLOCKTIME" in os.environ:
         kmp_blocktime = os.environ["KMP_BLOCKTIME"]
 
     # Set env
-    os.environ["KMP_AFFINITY"] = kmp_affinity
```
Comment (qiuxin2012) [inline pyzoo/zoo/common/nncontext.py]: delete these, same behavior with scala.

**Comment by qiyuangong** [comment]
@yangw1234 
Pls check python env part. :)

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 11]
```diff
@@ -1083,7 +1084,22 @@ private[zoo] class InternalDistriOptimizer[T: ClassTag] (
...
+     * And use multi OMP threads to speedup the single model's training.
+     * Currently, we only provide single model + multi OMP threads for torchnet model.
+     */
+    val torchNetOptimize = model.isInstanceOf[TorchNet]
+    val modelPerExecutor = if (torchNetOptimize) {
+      require(EngineRef.getEngineType() != MklDnn, "torchnet shouldn't use MKLDNN engine.")
+      val numOmpThread = distDataset.originRDD().sparkContext
+        .getConf.get("spark.executorEnv.OMP_NUM_THREADS").toInt
+      logger.info(s"torchnet will use ${numOmpThread} OMP threads.")
+      math.floor(EngineRef.getCoreNumber() / numOmpThread).toInt
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: I think you should just set it to 1; otherwise it can be 0 for Spark local

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 12]
```diff
@@ -1083,7 +1084,22 @@ private[zoo] class InternalDistriOptimizer[T: ClassTag] (
...
 
     val nodeNumber = EngineRef.getNodeNumber()
-    val coresPerNode = EngineRef.getCoreNumber()
+
+    /**
+     * The best practice of torchnet's training is single model in each executor.
+     * And use multi OMP threads to speedup the single model's training.
+     * Currently, we only provide single model + multi OMP threads for torchnet model.
+     */
+    val torchNetOptimize = model.isInstanceOf[TorchNet]
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: isTorchNet

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 13]
```diff
@@ -1125,9 +1141,16 @@ private[zoo] class InternalDistriOptimizer[T: ClassTag] (
 
       val modelsAndBroadcast = InternalOptimizerUtil.initThreadModels[T](
         trainingModel, distDataset, criterion, state,
-        Int.box(nodeNumber), Int.box(coresPerNode), Boolean.box(checkSingleton),
+        Int.box(nodeNumber), Int.box(modelPerExecutor), Boolean.box(checkSingleton),
         allReduceParameter, parameterSplits, validationMethods, optimMethods, parameterProcessors)
       cachedModels = modelsAndBroadcast._1
+      if (torchNetOptimize) {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: wrap this in another method `setMKLThread`

**Comment by jason-dai** [comment]
Please provide a summary of the expected behavior in the PR description.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala [hunk 14]
```diff
@@ -1089,14 +1096,15 @@ private[zoo] class InternalDistriOptimizer[T: ClassTag] (
      * The best practice of torchnet's training is single model in each executor.
      * And use multi OMP threads to speedup the single model's training.
      * Currently, we only provide single model + multi OMP threads for torchnet model.
+     * TODO: support tfnet.
      */
-    val torchNetOptimize = model.isInstanceOf[TorchNet]
+    val torchNetOptimize = TorchNet.isTorchNet(model)
     val modelPerExecutor = if (torchNetOptimize) {
       require(EngineRef.getEngineType() != MklDnn, "torchnet shouldn't use MKLDNN engine.")
       val numOmpThread = distDataset.originRDD().sparkContext
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: why do you still need numOmpThread?

Comment (qiuxin2012) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/models/Topology.scala]: To print a log below to notify user torch optimize is working.

**Comment by qiuxin2012** [comment]
jenkins passed
